src/service/promql/value.rs

```diff
@@ -97,49 +126,155 @@ impl Serialize for RangeValue {
+    /// Should only be used with counters.
+    ///
+    /// See <https://prometheus.io/docs/prometheus/latest/querying/functions/#rate>
+    Rate,
+
+    /// Calculate the increase in the time series. Adjust for counter resets.
+    /// Should only be used with counters.
+    ///
+    /// See <https://prometheus.io/docs/prometheus/latest/querying/functions/#increase>
+    Increase,
+
+    /// Calculate the difference between the first and last value. Don't adjust for counter resets. Should only be used with gauges.
+    ///
+    /// See <https://prometheus.io/docs/prometheus/latest/querying/functions/#delta>
+    Delta,
+}
+
+/// `extrapolated_rate` is a utility function for rate/increase/delta.
+///
+/// Calculates the rate (allowing for counter resets if `kind` is Rate or
+/// Increase), extrapolates if the first/last sample is close to the boundary,
+/// and returns the result as either per-second (if `kind` is Rate) or overall.
+///
+/// Returns `None` if there are fewer than two samples.
+///
+/// See the diagrams at <https://promlabs.com/blog/2021/01/29/how-exactly-does-promql-calculate-rates/#extrapolation-of-data>
+///
+/// # Panics
+///
+/// Panics if the samples are not in the range.
+//
+// cf. https://github.com/prometheus/prometheus/blob/80b7f73d267a812b3689321554aec637b75f468d/promql/functions.go#L67
+pub(crate) fn extrapolated_rate(
+    samples: &[Sample],
+    eval_ts: i64,
+    range: Duration,
+    offset: Duration,
+    kind: ExtrapolationKind,
+) -> Option<f64> {
+    let start = {
+        let range_plus_offset = range
+            .checked_add(offset)
+            .expect("BUG: overflow")
+            .as_micros()
+            .try_into()
+            .expect("BUG: integer conversion failed");
+        eval_ts
+            .checked_sub(range_plus_offset)
+            .expect("BUG: overflow")
+    };
+    assert!(start > 0);
+    let end = eval_ts
+        .checked_sub(
+            offset
+                .as_micros()
+                .try_into()
+                .expect("BUG: integer conversion failed"),
+        )
+        .expect("BUG: overflow");
+    assert!(end > 0);
+    assert!(start <= end);
+
+    if samples.len() < 2 {
+        // Not enough samples.
+        return None;
     }
+    let first = &samples[0];
+    let last = &samples.last().unwrap();
+
+    // The caller must ensure that the samples are in the range.
+    assert!(first.timestamp <= last.timestamp);
+    assert!(first.timestamp >= start);
+    assert!(last.timestamp <= end);
+
+    let mut result = last.value - first.value;
+
+    let is_counter = matches!(kind, ExtrapolationKind::Rate | ExtrapolationKind::Increase);
+    if is_counter {
+        // Handle counter resets.
+        let mut prev_value = first.value;
+        for sample in &samples[1..] {
+            if sample.value < prev_value {
+                result += prev_value;
+            }
+            prev_value = sample.value;
+        }
+    }
```

Comment:
We already handled conter reset when we load data from datafusion. Is those two places repeat calculate?